### PR TITLE
Added Moving Traps with 4 Slime Types

### PR DESCRIPTION
**Pull request by Pukami**
Major Changes:
- Added a "MobileTrap" base class for movable traps.
- Added a base "Trap" class that all traps inherit from. 
- Changed how client update is handled, giving a specific c_update call
- All game objects have update_state() called, except for tiles.

Minor Changes:
- Added a bunch of Slime assets
- Reset function for game objects that is called in beginning of build phase
- Player can pick up objects when invulnerable, stun length is longer though.
- Essence amount gained is only lost in multiples of 10. 

Needs Work:
- Yeah, the way Grid has a huge case statement for Constructs is ugly.
- Drop 20% essence on taking damage. This will be changed in the future probably.
- Value of each essence is 10, and is hardcoded on Player. not sure where that variable would be.

### CODE REVIEW

File: src/game_objects/construct.cpp [hunk 1]
```diff
@@ -8,6 +8,9 @@
 Construct::Construct(int x, int z, int id) : GameObject(id) {
     tag = "CONSTRUCT";
     set_position({ x, 0.0f, z });
+
+    initial_x = x;
+    initial_z = z;
```
Comment (jytang) [inline src/game_objects/construct.cpp]: make a copy of the transform (initial_transform)

**Review by AznCelebration — APPROVED** [review]

File: src/ui/ui_manager.cpp [hunk 2]
```diff
@@ -22,21 +22,12 @@ UIManager::UIManager(float aspect)
...
-        switch (type) {
-        case ConstructType::CHEST:
-            constructs.push_back(Constructs::CONSTRUCTS.at(ConstructType::CHEST));
-            break;
-        case ConstructType::SPIKES:
-            constructs.push_back(Constructs::CONSTRUCTS.at(ConstructType::SPIKES));
-            break;
-        case ConstructType::REMOVE:
+    for (int i = 1; i <= 12; i++) {
+        int num_available_constructs = 7; // Need better way than hardcoding this number
```
Comment (richard1) [inline src/ui/ui_manager.cpp]: 🎉  `Constructs::CONSTRUCTS.size();`

File: src/game/construct_types.h [hunk 3]
```diff
@@ -17,6 +17,10 @@ namespace Constructs {
     const static std::map<ConstructType, ConstructData> CONSTRUCTS = {
         { ConstructType::CHEST, { ConstructType::CHEST, 50, "Chest", "A treasure chest full of gold.", "test.png" } },
         { ConstructType::SPIKES, { ConstructType::SPIKES, 50, "Spikes", "It was me, spikes.", "dio.jpg" } },
+	    { ConstructType::SLIME_B, { ConstructType::SLIME_B, 10, "Blue Slime", "Likes to loop", "slime_blue.png" } },
```
Comment (richard1) [inline src/game/construct_types.h]: Seems to be a tab here

File: src/net/network_manager.cpp [hunk 4]
```diff
@@ -2,6 +2,7 @@
 #include "util/config.h"
 #include "game_objects/player.h"
 #include "game_objects/essence.h"
+#include "game_objects/traps/slime.h"
```
Comment (richard1) [inline src/net/network_manager.cpp]: Unneeded include?

**Review by richard1 — CHANGES_REQUESTED** [review]
Couple small things but LGTM

:goberserk:

**Pull request merged**